### PR TITLE
Make gap generators work when shroud is disabled.

### LINKED ISSUE #6123: RA1 Gap technology fails to work when Shroud is disabled
What I mean is that if Shroud is turned off in the game settings (most of my friends like to play with it off because seeing the terrain is a huge thing if you don't remember the map), then units/buildings like the Gap Generator or the vehicle that does the same thing don't work at all. We tried and it basically did nothing.

Could this be fixed in any way so that the Allies don't lose 2 units when one game setting is turned off? :O

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/3426756-ra1-gap-technology-fails-to-work-when-shroud-is-disabled?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by deniz1a**
Fixes #6123. This will look at game settings, if fog of war is disabled, gap gens will work like they do now. If fog of war is enabled, gap gens will generate fog and there won't be any frozen actors visible while they are operational. Also this commit disables auto targeting for units/structures under gap, even if you have satellite. So satellite icons will also not appear where gap gen is active.

I tested this and seems to work.

### CODE REVIEW

File: OpenRA.Game/Orders/UnitOrderGenerator.cs [hunk 1]
```diff
@@ -74,7 +74,7 @@ public string GetCursor(World world, CPos xy, MouseInput mi)
 			else
 			{
 				var frozen = world.ScreenMap.FrozenActorsAt(world.RenderPlayer, mi)
-					.Where(a => a.Info.Traits.Contains<ITargetableInfo>() && !a.Footprint.All(world.ShroudObscures))
+					.Where(a => a.Info.Traits.Contains<ITargetableInfo>() && !a.Footprint.All(world.ShroudObscures) && !world.RenderPlayer.Shroud.IsUnderGapGen(a.Footprint))
```
Comment (RoosterDragon) [inline OpenRA.Game/Orders/UnitOrderGenerator.cs]: This code is duplicated in the Order method, will that need updating too?

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs [hunk 2]
```diff
@@ -119,8 +119,8 @@ public FrozenActorLayer(Actor self)
 
 		public void Add(FrozenActor fa)
 		{
-			frozen.Add(fa.ID, fa);
-			world.ScreenMap.Add(owner, fa);
+                frozen.Add(fa.ID, fa);
+                world.ScreenMap.Add(owner, fa);
```
Comment (RoosterDragon) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: Use tabs for indentation.

File: OpenRA.Game/Traits/World/Shroud.cs [hunk 3]
```diff
@@ -28,7 +28,7 @@ public class Shroud
 		readonly Map map;
 
 		readonly CellLayer<short> visibleCount;
-		readonly CellLayer<short> generatedShroudCount;
+		public readonly CellLayer<short> generatedGapCount;
```
Comment (RoosterDragon) [inline OpenRA.Game/Traits/World/Shroud.cs]: You can keep this private.

File: OpenRA.Game/Traits/World/Shroud.cs [hunk 4]
```diff
@@ -275,7 +275,10 @@ public bool IsExplored(MPos uv)
 
 		bool IsExploredCore(MPos uv)
 		{
-			return explored[uv] && (generatedShroudCount[uv] == 0 || visibleCount[uv] > 0);
+            if (FogEnabled)
+                return explored[uv];
+            else
+                return explored[uv] && (generatedGapCount[uv] == 0 || visibleCount[uv] > 0);
```
Comment (RoosterDragon) [inline OpenRA.Game/Traits/World/Shroud.cs]: ```return explored[uv] && (FogEnabled || generatedGapCount[uv] == 0 || visibleCount[uv] > 0);``` would be equivalent, also check indentation here.

File: OpenRA.Game/Orders/UnitOrderGenerator.cs (hunk 1 above)
Comment (deniz1a) [inline OpenRA.Game/Orders/UnitOrderGenerator.cs]: I don't know, it seems to work with just this code. This is to prevent having an attack cursor for hidden buildings under fog (they dont have frozen images now if gap gen is active).

File: OpenRA.Game/Traits/World/Shroud.cs [hunk 5]
```diff
@@ -359,5 +372,21 @@ public bool HasFogVisibility()
 		{
 			return fogVisibilities.Value.Any(f => f.HasFogVisibility(self.Owner));
 		}
+        
+        public bool IsUnderGapGen(MPos[] fp)
+        {
+            if (fp.All(c => generatedGapCount[c] > 0))
```
Comment (RoosterDragon) [inline OpenRA.Game/Traits/World/Shroud.cs]: You can do ```return fp.All(c => generatedGapCount[c] > 0);```. Check indentation.

File: OpenRA.Game/Traits/World/Shroud.cs [hunk 6]
```diff
@@ -359,5 +372,21 @@ public bool HasFogVisibility()
...
+                return true;
+            else
+                return false;
+        }
+        
+        public bool IsUnderGapGen(IEnumerable<CPos> fp)
```
Comment (RoosterDragon) [inline OpenRA.Game/Traits/World/Shroud.cs]: You can do ```return fp.All(c => generatedGapCount[c] > 0);```. Check indentation.

File: OpenRA.Mods.RA/Effects/GpsDot.cs [hunk 7]
```diff
@@ -65,6 +65,9 @@ public GpsDot(Actor self, GpsDotInfo info)
 
 		bool ShouldShowIndicator()
 		{
+            if(self.World.RenderPlayer.Shroud.IsUnderGapGen(Shroud.GetVisOrigins(self)))
```
Comment (RoosterDragon) [inline OpenRA.Mods.RA/Effects/GpsDot.cs]: Check indentation and leave a space after the ```if``` keyword.

**Comment by deniz1a** [comment]
OK I think I fixed all code styling issues. I also squashed all commits.

**Comment by deniz1a** [comment]
Also for gap gen to cover a building, all its footprint cells must be within range. So if part of the building is outside gap gen range it will appear as normal. Does footprint include bibs, the space below structures?

**Comment by pchote** [comment]
Shroud is an engine-level feature, so please don't introduce RA-specific logic and naming there.  You can do that instead in the mod-level traits.

**Comment by deniz1a** [comment]
Do you mean UpdateShroudGeneration to UpdateGapGeneration etc changes? But it doesn't generate shroud anymore if fog is enabled. Isn't gap a general name for it?

**Comment by Phrohdoh** [comment]
No, shroud is the name for shroud and fog is the name for fog.

**Comment by deniz1a** [comment]
OK so how should it be called? It's not a shroud generator or fog generator but a combination of those. ShroudFogGeneration?

**Comment by deniz1a** [comment]
Actually if fog is enabled, gap generator doesn't generate fog. The fog is generated automatically by the engine. What gap gen does then is remove frozen actors from fog. So it either generates shroud or removes frozen actors from fog. WHat should it be called? Gap generator seems to be the most appropriate name, it generates a gap in enemy radar/map. Are there any other name suggestions? ShroudFogGeneration doesn't look suitable since gap gen doesn't generate fog.

**Comment by deniz1a** [comment]
Also the other point pchote made, I haven't introduced any ra specific code here? The shroud generation code was already there, I just changed its behaviour. And don't all mods have shroud and fog option? So it's not mod specific.

**Comment by pchote** [comment]
The concept of having the "shroud generation" do one thing if shroud is enabled but another if it isn't is a RA specific idea.  Other mods may want to have pure shroud generators or pure cloak generators, but you are instead enforcing them to do what the gap generator does.

The shroud code needs to keep these ideas separate (with separate methods and state for shroud and cloak generation), and then the gameplay code (i.e. CreatesShroud) should expose trait properties so that the generation type can be configured in the yaml.

**Comment by deniz1a** [comment]
OK it was agreed in IRC to call this functionality "Stealth Field Generation". I will make the necessary changes and send another commit.

**Comment by deniz1a** [comment]
I decided to postpone this until #7577 is merged and #7669 is fixed.

**Comment by deniz1a** [comment]
OK I have another naming question: So the general trait will be called CreatesStealth, it is a general stealth generation field. And then there will be two types of stealth. One is the current gap generator, it will generate shroud, so it's called type: shroud. But what should the new type be called? The new gap generator behaviour will act as frozen actor remover from fog and also prevent units from being auto targeted. Sensor disruptor was suggested in IRC. What do you think?

**Comment by Phrohdoh** [comment]
:-1: for `CreatesStealth`, shroud is not stealth/cloak.

**Comment by penev92** [comment]
This is basically what I've gathered from all the discussions:
 - Working with Shroud disabled - it generates fog
 - Working with Fog disabled - it generates shroud
 - Working with both Shroud and Fog disabled - it just makes things "invisible" unless you are near them
 - Working with both Shroud and Fog enabled - generate fog

**Comment by deniz1a** [comment]
Yes. Or simpler: if fog is enabled, gap gen acts as disruption generator. If fog is disabled but shroud is enabled, it generates shroud. If both are disabled, it makes units invisible?

**Comment by deniz1a** [comment]
@Phrohdoh: So what should it be called instead of CreatesStealth.cs?

File: OpenRA.Game/OpenRA.Game.csproj [hunk 8]
```diff
@@ -349,4 +349,4 @@
   </Target>
   -->
   <ItemGroup />
-</Project>
+</Project>
```
Comment (Phrohdoh) [inline OpenRA.Game/OpenRA.Game.csproj]: This needs to be reverted.

**Comment by deniz1a** [comment]
OK I have updated the code. Another review please.

File: OpenRA.Game/OpenRA.Game.csproj (hunk 8 above)
Comment (deniz1a) [inline OpenRA.Game/OpenRA.Game.csproj]: What needs to be reverted? Those changes in the project file were done by Visual Studio automatically when I renamed CreatesShroud.cs to CreatesConcealment.cs.

**Comment by deniz1a** [comment]
I didn't add separate methods for disruption field generation, I couldn't decide which one would be better. If you say separate methods for disruption in Shroud.cs would be better, I can change it. And I made a single trait for both types. I think this is better than having two traits and then checking if the actor also has the other trait... Because they are basically the same trait with slightly different behaviour.

File: OpenRA.Game/Traits/World/Shroud.cs [hunk 9]
```diff
@@ -30,15 +30,17 @@ public class Shroud
 		readonly Map map;
 
 		readonly CellLayer<short> visibleCount;
-		readonly CellLayer<short> generatedShroudCount;
 		readonly CellLayer<bool> explored;
+		readonly CellLayer<short> generatedDisruptionCount;
+		readonly CellLayer<short> generatedShroudCount;
```
Comment (deniz1a) [inline OpenRA.Game/Traits/World/Shroud.cs]: It doesn't work when I make this CellLayer<ushort>. Why? Does it ever fall to negative values?

**Comment by deniz1a** [comment]
I have finished the code. It adds a new trait called CreatesDisruptionField. This makes gap gens work without shroud.

File: OpenRA.Game/Traits/World/Shroud.cs (hunk 9 above)
Comment (Phrohdoh) [inline OpenRA.Game/Traits/World/Shroud.cs]: An unsigned type can **never** be negative.

Comment (deniz1a) [inline OpenRA.Game/Traits/World/Shroud.cs]: I know. But does it fall to negative when it's short?

**Comment by Phrohdoh** [comment]
This certainly *works*, but as I've stated before I'm not an advocate for this logic.

I do not think this is visually pleasing: (structures disappearing into the fog abruptly)
gif

**Comment by deniz1a** [comment]
OK but how could it be done otherwise so that gap generators work when shroud is disabled?

File: OpenRA.Game/Traits/CreatesDisruptionField.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+ */
+#endregion
+
+using System.Linq;
+
+namespace OpenRA.Traits
+{
+	public class CreatesDisruptionFieldInfo : ITraitInfo
+	{
+		public readonly WRange Range = WRange.Zero;
```
Comment (Phrohdoh) [inline OpenRA.Game/Traits/CreatesDisruptionField.cs]: Add a `Desc` attribute to this, please.

File: OpenRA.Game/Traits/CreatesDisruptionField.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Linq;
+
+namespace OpenRA.Traits
+{
+	public class CreatesDisruptionFieldInfo : ITraitInfo
```
Comment (Phrohdoh) [inline OpenRA.Game/Traits/CreatesDisruptionField.cs]: Add a `Desc` attribute to this, please.

File: OpenRA.Game/Traits/CreatesDisruptionField.cs [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+			var disabled = self.TraitsImplementing<IDisable>().Any(d => d.Disabled);
+			if (cachedLocation != self.Location || cachedDisabled != disabled)
+			{
+				cachedLocation = self.Location;
+				cachedDisabled = disabled;
+				Shroud.UpdateDisruptionGenerator(self.World.Players.Select(p => p.Shroud), self);
+			}
+		}
+
+		public WRange Range { get { return cachedDisabled ? WRange.Zero : info.Range; } }
```
Comment (Phrohdoh) [inline OpenRA.Game/Traits/CreatesDisruptionField.cs]: This should go above all methods and private members.

File: OpenRA.Game/Traits/World/Shroud.cs (hunk 9 above)
Comment (Phrohdoh) [inline OpenRA.Game/Traits/World/Shroud.cs]: You tell us.
Add some debug lines and you'll figure it out.

File: OpenRA.Game/Traits/World/Shroud.cs [hunk 13]
```diff
@@ -163,42 +166,105 @@ void UpdateVisibility(Actor a, ref CPos[] visible)
...
+				return;
+
+			// Lazily generate the disrupted tiles, allowing the caller to re-use them if desired.
+			underDisruption = underDisruption ?? FindVisibleTiles(a, cd.Range);
+
+			foreach (var c in underDisruption)
+				generatedDisruptionCount[c]++;
+
+			if (disruptionGenerators.ContainsKey(a))
+				throw new InvalidOperationException("Attempting to add duplicate shroud generation");
```
Comment (Phrohdoh) [inline OpenRA.Game/Traits/World/Shroud.cs]: If you are going to change logic please update any accompanying messages.
This is now *disruption*, not shroud.

**Comment by deniz1a** [comment]
This code is still not finished. I found some bugs and am trying to fix them...

**Comment by deniz1a** [comment]
This patch has turned out to be more complex than I had thought. Currently it has the following problems:
1) If a previously discovered building that is under active gap gen (disruption field) is sold or destroyed, its frozen actor doesn't reappear if the gap gen is not active anymore. In other words, the information that it is no longer there is revealed, which is not done for normal frozen actors.
2) Whether a frozen actor is rendered or not is decided by using its footprint. But whether it can be auto-targeted is decided by its OccupiedCells(). These are different areas, which results in buildings that are visible as frozen actors but not auto-targeted, if the occupied cells are under disruption field but the bibs of the building (which are part of the footprint) are not.
3) When a move order is given over a hidden frozen actor, it is registered as attack order.
4) GPS dot appears for a moment over frozen actors sometimes.

I am trying to fix these issues but I'm having a hard time. Any help is appreciated.

**Comment by deniz1a** [comment]
I decided to postpone this PR again. I am stuck at fixing issue #1 described above... pchote said there may be some changes to how frozen actors are rendered. So I am waiting either that to be finished or someone to show me how I could fix these problems.

**Comment by deniz1a** [comment]
I have fixed the three bugs described above. GPS icon appearing briefly over newly rendered frozen actors is not caused by this PR, it is also present in bleed. This PR is now finished. I tested it and couldn't find any other issues.

File: OpenRA.Game/OpenRA.Game.csproj (hunk 8 above)
Comment (RoosterDragon) [inline OpenRA.Game/OpenRA.Game.csproj]: You should stage just the ```<Compile Include="Traits\CreatesDisruptionField.cs" />``` line. We're trying to avoid staging other lines the different IDEs keep automatically changing so they don't go back and forth all the time.

File: OpenRA.Game/Traits/CreatesDisruptionField.cs [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+			this.info = info;
+			lobbyFogDisabled = !self.World.LobbyInfo.GlobalSettings.Fog;
+		}
+
+		public void Tick(Actor self)
+		{
+			if (lobbyFogDisabled)
+				return;
+
+			var disabled = self.TraitsImplementing<IDisable>().Any(d => d.Disabled);
```
Comment (RoosterDragon) [inline OpenRA.Game/Traits/CreatesDisruptionField.cs]: ```var disabled = self.IsDisabled();```

File: OpenRA.Game/Traits/World/Shroud.cs [hunk 15]
```diff
@@ -163,42 +162,105 @@ void UpdateVisibility(Actor a, ref CPos[] visible)
 			AddVisibility(a, ref visible);
 		}
 
-		void AddShroudGeneration(Actor a, ref CPos[] shrouded)
+		void AddDisruptionGenerator(Actor a, ref CPos[] underDisruption)
+		{
+			var cd = a.TraitOrDefault<CreatesDisruptionField>();
+			if (cd == null || a.Owner.IsAlliedWith(self.Owner) || cd.Range == WRange.Zero || !self.World.LobbyInfo.GlobalSettings.Fog)
```
Comment (RoosterDragon) [inline OpenRA.Game/Traits/World/Shroud.cs]: Isn't the check for the disabled fog also done at the trait level? Do we need to duplicate it here?

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs [hunk 16]
```diff
@@ -179,7 +186,27 @@ public void Tick(Actor self)
 		public virtual IEnumerable<IRenderable> Render(Actor self, WorldRenderer wr)
 		{
 			return world.ScreenMap.FrozenActorsInBox(owner, wr.Viewport.TopLeft, wr.Viewport.BottomRight)
-				.Where(f => f.Visible)
+				.Where(f =>
```
Comment (RoosterDragon) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: Could you walk me through what this logic block is trying to solve? You've clearly engineered it very carefully to fix some of the issues you were seeing but it's difficult to see why each of these cases is needed. I'm wondering if there's something we could do to simply this concept.

Comment (deniz1a) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: OK. I made that to fix the bug #1 described above. The frozen actor of the building has to be rendered at least once so that it is cached and is shown even when the building is not there anymore. This code makes sure that every frozen actor is rendered once even if it's under disruption field and then one tick passes while it is rendered. Because tick and render frequencies are different. This makes the renderables cached so that they can be shown again when the building is destroyed and disruption field is gone.

At least this is my understanding of how the engine works. Is that correct?

**Comment by deniz1a** [comment]
I made the changes you requested. While I was doing that I also realized that there should also be a separate RenderDisruptionCircle trait to draw the range circle. And shouldn't this and RenderShroudCircle be in Mods.Common? Or even in the engine itself. Because they are part of CreatesShroud and CreatesDisruptionField traits.

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs (hunk 16 above)
Comment (deniz1a) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: OK, I unnestified the if statements. I don't think it can be simplified more than this.

Comment (pchote) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: This sounds much more complicated than it needs to be.  A simpler solution would be to change FrozenActor to generate the sprites from an ActorPreview if the actor has been destroyed.

Comment (deniz1a) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: How can I do that?

**Comment by deniz1a** [comment]
And the developing continues...
[19:54:56] <+pchote> find the thing that calls the render, and then create an actor preview
[19:55:03] <+pchote> and then render that and stash the sprites
[19:55:46] <+pchote> actually, there's a problem there
[19:55:59] <+pchote> the base ActorPreview class currently lives in Mods.Common
[19:56:07] <+pchote> it will need to move into game for you to be able to do this

**Comment by deniz1a** [comment]
I have changed this to use ActorPreview for rendering the frozen actor. I also added some checks so that it is only rendered when the building state has changed.

File: OpenRA.Game/OpenRA.Game.csproj [hunk 17]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (abcdefg30) [inline OpenRA.Game/OpenRA.Game.csproj]: This needs to be undone.

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs [hunk 18]
```diff
@@ -25,6 +26,7 @@ public class FrozenActorLayerInfo : ITraitInfo
 	public class FrozenActor
 	{
 		public readonly MPos[] Footprint;
+		public readonly CPos[] OccCells;
```
Comment (Phrohdoh) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: What is `Occ`, Occupied?

Comment (deniz1a) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: Yes, that's for Occupied Cells. But there already is a method called OccupiedCells() so I wanted to give this a different name to avoid confusion.

Comment (Phrohdoh) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: This sounds like duplication then.

Comment (deniz1a) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: But that method cannot be used if the actor is null. A frozen actor can exist even after its actor is destroyed. The actor's occupied cells have to be stored in the frozen actor so that they can be used even if the actor is null.

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs [hunk 19]
```diff
@@ -48,6 +48,7 @@ public FrozenActor(Actor self, MPos[] footprint, CellRegion footprintRegion, Shr
 			isVisibleTest = shroud.IsVisibleTest(footprintRegion);
 
 			Footprint = footprint;
+			OccCells = Shroud.GetVisOrigins(actor).ToArray();
```
Comment (deniz1a) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: Should I use "OccCells = self.OccupiesSpace.OccupiedCells().Select(c => c.First).ToArray();" instead of this line? Do I need the checks in GetVisOrigins()?

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs (hunk 18 above)
Comment (pchote) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: That is what the Footprint above is for. Use that instead.

File: OpenRA.Game/Traits/World/Shroud.cs [hunk 20]
```diff
@@ -383,13 +453,25 @@ public bool IsVisible(Actor a)
 
 		public bool IsTargetable(Actor a)
 		{
+			IEnumerable<CPos> vis = GetVisOrigins(a);
+			bool concealed = vis.All(c => generatedDisruptionCount[c] > 0 || generatedShroudCount[c] > 0);
```
Comment (deniz1a) [inline OpenRA.Game/Traits/World/Shroud.cs]: pchote: But I'm using occupied cells here to determine if a building is targetable. Should I use footprint here too? I have to use the same condition everywhere.

File: OpenRA.Game/Orders/UnitOrderGenerator.cs [hunk 21]
```diff
@@ -29,7 +29,8 @@ public IEnumerable<Order> Order(World world, CPos xy, MouseInput mi)
 			else
 			{
 				var frozen = world.ScreenMap.FrozenActorsAt(world.RenderPlayer, mi)
-					.Where(a => a.Info.Traits.Contains<ITargetableInfo>() && !a.Footprint.All(world.ShroudObscures))
+					.Where(a => a.Info.Traits.Contains<ITargetableInfo>() && !a.Footprint.All(world.ShroudObscures)
+						&& !world.RenderPlayer.Shroud.IsUnderDisruptionField(a.OccCells))
```
Comment (pchote) [inline OpenRA.Game/Orders/UnitOrderGenerator.cs]: Requiring an explicit check for IsUnderDisruptionField here is a sign that the existing code is bogus.  This code (from before your changes) should have been filtering out FrozenActors that aren't visible, and your changes should have just worked (because it should be making FrozenActors.Visible -> false).  Please fix this properly by adding a filter on Visible instead of hacking in a disruption check in these places.

File: OpenRA.Game/Traits/CreatesShroud.cs [hunk 22]
```diff
@@ -21,20 +22,21 @@ public class CreatesShroudInfo : ITraitInfo
 
 	public class CreatesShroud : ITick, ISync
 	{
+		public WRange Range { get { return cachedDisabled ? WRange.Zero : info.Range; } }
```
Comment (pchote) [inline OpenRA.Game/Traits/CreatesShroud.cs]: Why did you move this line?

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs [hunk 23]
```diff
@@ -64,6 +65,8 @@ public FrozenActor(Actor self, MPos[] footprint, CellRegion footprintRegion, Shr
 		int flashTicks;
 		IRenderable[] renderables = NoRenderables;
 		bool needRenderables;
+		DamageState DamageStateCached;
```
Comment (pchote) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: These are private, so should the naming needs to be camelCase.

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs [hunk 24]
```diff
@@ -98,23 +101,22 @@ public void Flash()
...
+			if (needRenderables && (renderables == NoRenderables || DamageStateCached != DamageState || OwnerCached != Owner))
 			{
-				needRenderables = false;
-				if (!actor.Destroyed)
-				{
-					IsRendering = true;
-					renderables = actor.Render(wr).ToArray();
-					IsRendering = false;
-				}
+				IsRendering = true;
```
Comment (pchote) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: Please only use the ActorPreview if `actor.Destroyed` is true.  Using it for a living actor breaks the feature where the actor is frozen with its last visible frame (including muzzle flashes, turret position, etc).

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs [hunk 25]
```diff
@@ -98,23 +101,22 @@ public void Flash()
 
 		public IEnumerable<IRenderable> Render(WorldRenderer wr)
 		{
-			if (needRenderables)
+			if (needRenderables && (renderables == NoRenderables || DamageStateCached != DamageState || OwnerCached != Owner))
```
Comment (pchote) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: These extra checks don't look right to me.  if needRenderables is true, then I would expect it to generate renderables.  If there is some condition where you don't want it to generate renderables, then those conditions should be applied at the point that variable is set.

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs [hunk 26]
```diff
@@ -179,7 +181,7 @@ public void Tick(Actor self)
 		public virtual IEnumerable<IRenderable> Render(Actor self, WorldRenderer wr)
 		{
 			return world.ScreenMap.FrozenActorsInBox(owner, wr.Viewport.TopLeft, wr.Viewport.BottomRight)
-				.Where(f => f.Visible)
+				.Where(f => f.Visible && !owner.Shroud.IsUnderDisruptionField(f.OccCells))
```
Comment (pchote) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: Like I mentioned above, your disruption checks should be built into `f.Visible` instead of adding IsUnderDisruptionField everywhere.

File: OpenRA.Game/Traits/World/Shroud.cs (hunk 20 above)
Comment (pchote) [inline OpenRA.Game/Traits/World/Shroud.cs]: This code deals with regular actors (not FrozenActors), so I'm not sure what you're asking.

Comment (deniz1a) [inline OpenRA.Game/Traits/World/Shroud.cs]: But the checks for regular actors and frozen actors have to be the same. Otherwise there are frozen actors that are visible but not auto-targetable. So either I have to check for footprints here or add the occupied cells to frozen actor.

Comment (pchote) [inline OpenRA.Game/Traits/World/Shroud.cs]: If this is for targeting, then your "occupied cells" in the frozen actor need to instead store and deal with the targetable positions returned by the actor.  These are conceptually different to the occupied cells, even though they may (for the current set of traits) be derived from the same data.

Comment (pchote) [inline OpenRA.Game/Traits/World/Shroud.cs]: Although i'm still not convinced about all of this.  It sounds like you have uncovered an existing bug that must already exist in certain situations in the current code -- separating out the bugfix from your generation code would make this much easier to understand and review.

File: OpenRA.Game/Traits/CreatesShroud.cs (hunk 22 above)
Comment (deniz1a) [inline OpenRA.Game/Traits/CreatesShroud.cs]: Because Phrohdoh told me to.

Comment (penev92) [inline OpenRA.Game/Traits/CreatesShroud.cs]: He does have a point. Properties usually go before methods.

**Comment by deniz1a** [comment]
This PR depends on #8084.

**Comment by penev92** [comment]
I am closing this in an attempt to clean up the PR queue a bit. Feel free to reopen it once the conceptual problems are cleared up.

**Comment by deniz1a** [comment]
This apparently depends on satellite code being fixed to handle visibility and stuff. Related: #8203
